### ISSUE TITLE
Epic: Event hooks for executing Jobs - JobHooks

### ISSUE THREAD

**Issue by lampwins** [nautobot/nautobot#862]
### Environment
* Python version: 3.9
* Nautobot version: 1.1.2

### Proposed Functionality
As Austin the Network Automation Engineer, I want data change events in Nautobot to be able to trigger defined Jobs for execution so that I can build more complex event processing logic within Nuatobot that does not fit the webhook use case.

### Use Case
Today nautobot supports webhooks that fire on create, update, and delete events tied to specific models. The consumption of these webhook events requires a separate system which makes certain types of event flows overly complex. Notably, this use case speaks to follow-up actions within Nautobot in response to a change in data. For instance, I might want the next IP address in a pool to be automatically assigned to an interface that just had a circuit termination connected. Or I could have all of the Virtual Machines on a given cluster moved to another cluster if the cluster's hosts have been set to a status of offline.

The point is I have complex business logic that I want to execute within Nautobot Jobs, in the same way, I work with webhooks, in cases where webhooks make this automation unnecessarily complex.

I expect all the same context currently available in a webhook to be provided as input to a Job I define.

### Database Changes
We will need a new model similar to Webhook for the definition of these "JobHooks"

### External Dependencies
There should be no external dependencies to implement this with existing functionality.

### Requirements

- [ ] Limit to just data changes (same as webhooks)
- [ ] Mapping data changes to job inputs
- [ ] Job execution should track the source of the execution (UI, API, Object Change event, or Job Hook)
- [ ] Job Hooks may not be executed by other Job Hooks

### Out of Scope
- [ ] Chaining of jobs by the way of assigning a JobHook to the completion of another Job
- [ ] Replacing webhooks with this pattern

### Stories
- [x] #1876
- [x] #1877
- [x] #1878

**Comment by glennmatthews** [comment]
Need to think about how to mesh the Webhook data pattern (pre/post change dicts, `obj`) with the Job `Var` pattern. Do we need a new Var type (for a dict, perhaps?) Would this Var type be restricted to only JobHook-able Jobs and not for general-use Jobs invokable by the user? Do we need the ability for the JobHook definition to specify values for other Vars (StringVar, IPAddressVar, etc.) as inputs to a more traditional job? Maybe even templating-style (supply `obj.ip_address` as input to a given IPAddressVar)?

**Comment by lampwins** [comment]
Let's create a spike around investigation of potential issues related to celery task execution.

**Comment by lampwins** [comment]
Blocked by #1292

**Comment by lampwins** [comment]
Closing this out as complete!
